### PR TITLE
[mono][debugger] Fix watch of local variable values

### LINKED ISSUE #6161: Android debugger showing variables wrongly (.Net6)
Android debugger is showing incorrect values, either by hovering the variables in code or using "Add watch" or "Quick watch"

As it can be seen in the below video, the "fontFamily" variable is showing true, as if it were a boolean variable, and the "typeface" variable is showing a string that has the value of the "fontFamily" variable. Similarly textStyle shows the value of the "a" variable.

android debugger bug

### PR DESCRIPTION
**Pull request by thaystg**
Fixes https://github.com/xamarin/xamarin-android/issues/6161

This was a side-effect of icordbg support implementation.

In this case the index that come from debugger-libs is 0 but the correct index after reading pdb info is 1. That is why it was showing wrong variable value.

The index that comes from icordbg is already the correct one read from pdb we don't need to "fix" it on runtime side.

### CODE REVIEW

**Comment by lewing** [comment]
is it only the 0 index that is off?

**Review by lambdageek — DISMISSED** [review]

**Comment by thaystg** [comment]
No, if you have more than one local variable all of them will be wrong:
like the 0 -> would take the real 0 positioned -> should take the 1 positioned.
like the 1 -> would take the real 1 positioned -> should take the 2 positioned.

It doesn't happen in all methods. I have tested in a lot of cases and couldn't reproduce, but today I found a code attached https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1366070/ and could reproduce.

The code sample with variable `view` showing wrong value.
```
set
            {
                someValue = value;
                count++;

                if (count == 10)
                {
#if __ANDROID__
                    var view = Handler.NativeView as Android.Views.View;

                    if (view != null)
                    {

                    } // Add your break point here and you'll see the value of view is "true"
#endif
                }

                SomeValue = count;
            }
```

**Comment by lambdageek** [comment]
1. Is it possible to add a test for this?

2. Should we try to get this into RC1? or is it ok to wait for RC2?

File: src/mono/mono/component/debugger-agent.c [hunk 1]
```diff
@@ -9338,8 +9338,8 @@ frame_commands (int command, guint8 *p, guint8 *end, Buffer *buf)
 			} else {
 				MonoDebugLocalsInfo *locals;
 
-				locals = mono_debug_lookup_locals (frame->de.method, TRUE);
-				if (locals) {
+				locals = mono_debug_lookup_locals (frame->de.method);
+				if (locals && CHECK_ICORDBG (FALSE)) {
 					g_assert (pos < locals->num_locals);
 					pos = locals->locals [pos].index;
 					mono_debug_free_locals (locals);
```
Comment (lewing) [inline src/mono/mono/component/debugger-agent.c:9345]: will this leak locals != null and it is icordbg?

**Comment by thaystg** [comment]
1) I'm not sure. I'm thinking. And also I'll test if this will not break wasm debugger now that it also uses debugger-agent.
2) I think we should backport.

File: src/mono/mono/component/debugger-agent.c (hunk 1 above)
Comment (thaystg) [reply src/mono/mono/component/debugger-agent.c:9351]: Yes. fixed.

**Comment by lewing** [comment]
so debugger libs is sending the 0-index index into the number of locals and icordbg is sending the index in the debug info?  Is it ever off by anything other than 1?

**Comment by thaystg** [comment]
Here is where we read the correct index from pdb: https://github.com/dotnet/runtime/blob/3ffd963a5b371a9fd9f30044e0c41bf295d1d4cd/src/mono/mono/metadata/debug-mono-ppdb.c#L701

In some methods the variable 0 has index 0.
In this specific case the variable 0 has index 1.

And I don't know the reason.

**Comment by thaystg** [comment]
This PR is probably breaking webassembly that also send the correct index.

**Comment by radical** [comment]
> Here is where we read the correct index from pdb:
> 
> https://github.com/dotnet/runtime/blob/3ffd963a5b371a9fd9f30044e0c41bf295d1d4cd/src/mono/mono/metadata/debug-mono-ppdb.c#L701
> 
> In some methods the variable 0 has index 0.
> In this specific case the variable 0 has index 1.
> 
> And I don't know the reason.

Not a static vs instance, thing, I guess?

**Review by lewing — APPROVED** [review]
👍

File: src/mono/mono/metadata/mono-debug.c [hunk 2]
```diff
@@ -893,18 +893,16 @@ mono_debug_lookup_locals (MonoMethod *method, mono_bool ignore_pdb)
...
-				res = mono_debug_symfile_lookup_locals (minfo);
-		}
+
+	if (minfo->handle->ppdb) {
+		res = mono_ppdb_lookup_locals (minfo);
+	} else {
+		if (!minfo->handle->symfile || !mono_debug_symfile_is_loaded (minfo->handle->symfile))
+			res = NULL;
+		else
+			res = mono_debug_symfile_lookup_locals (minfo);
```
Comment (radical) [inline src/mono/mono/metadata/mono-debug.c:903]: This is handling the case of embedded pdb, and pdb as a file, correct? Can the tests be updated to check for both?

Comment (thaystg) [reply src/mono/mono/metadata/mono-debug.c:903]: We cannot test it for wasm because wasm debugger will not use this code path.

Comment (lewing) [reply src/mono/mono/metadata/mono-debug.c:903]: not needed, it isn't a good symbol reading issue.

**Comment by radical** [comment]
Just for my understanding, this is a difference between debugger-libs+debugger-agent, and icordbg+debugger-agent? And wasm uses the latter, and what uses the former? Please correct me, if I have that wrong!

**Comment by lewing** [comment]
> Just for my understanding, this is a difference between debugger-libs+debugger-agent, and icordbg+debugger-agent? And wasm uses the latter, and what uses the former? Please correct me, if I have that wrong!

In the older protocol we referenced the locals differently (always 0 indexed).  That changed with icordbg support to match the index from the debug info but it wasn't noticed. This just makes that change explicit by marking it as a protocol change and handling the old behavior.

**Review by radical — APPROVED** [review]

**Pull request merged**